### ISSUE TITLE
Unable to run multiple pre-process-forwarder without event collision (dedup warning)

### ISSUE THREAD

**Issue by Sawk** [getsentry/onpremise#671]
## Version Information
Version: Sentry 20.8.017e488c

## Description
Running multiple pre-process-forwarder leads de-duplication warning as ingesters try to process the same event.

## Steps to Reproduce
1. Run multiple ingest consumer (run ingest-consumer --all-consumer-types)
2. Constantly send event (~10 event per seconds)

## Logs
07:09:26 [WARNING] sentry.ingest.ingest_consumer: pre-process-forwarder detected a duplicated event with id:6b07bbdf80f24e53ba4efdbce2029dc3 for project:8.

I have confirmed the ingest consumer have different consumer ID and are part of the same group ID on Kafka side.
I have digged into the source code and found this related part
https://github.com/getsentry/sentry/blob/master/src/sentry/utils/batching_kafka_consumer.py#L66
To my understanding this implementation will do some in-memory processing and not commit or flush (no auto-commit) to kafka until it has finished processing which will allow multiple ingest consumer to process the same event. 

This is then stopped at the logic level (hence the warning) with some dedup here
https://github.com/getsentry/sentry/blob/3aa5fd94bac7361ce9a766468c04a559b1ca7119/src/sentry/ingest/ingest_consumer.py#L127

The issue we might run to unless I am wrong is that with that approach I am not sure this can scale very well as increasing the consumer will just lead to having more collision trying to process the same event. Furthermore multi-threading (concurrency) has been removed: https://github.com/getsentry/sentry/pull/20356

**Comment by beezz** [comment]
Hi @Sawk !

As you confirmed, the consumers are in the same group ID different consumer ID meaning they are consuming from a different set of Kafka partitions and there should be no repeated processing for events.

Duplicates are usually the result of the consumer crashing or being restarted before it is able to commit the batch it is processing.

Also, it can happen shortly after you add new consumers and Kafka does the rebalance of consumers and partitions.

In your case, do you see the warnings consistently even after some time you added another consumer(s)?

**Comment by Sawk** [comment]
Hi @beezz 
Thank you for the quick reply, much appreciated.

**TLDR;** 
Kafka 2.3.0 has rebalancing bug, do not use especially if you adjust consumer count.

**Details**
I have had the time to dig a bit more on the issue and indeed the rebalancing should be occurring and assign new consumer ID within the same group to the partition. Rebalancing should take seconds or minutes but seemed it was still hitting the de-dup warning.

When checking more in details, the rebalancing seemed to have a weird behavior and loop-rebalancing all the time. Deleting the topic and redeploying seemed as well to be working, so the bug seemed to be more on Kafka side.

The Kafka version (2.3.0) that we use, seems to have a quite huge bug in that regard, that has been fixed in 2.3.1
https://issues.apache.org/jira/browse/KAFKA-8653
https://downloads.apache.org/kafka/2.3.1/RELEASE_NOTES.html
Having the same amount of consumer all the time from the start do not trigger the issue, however when scaling (which will cause rebalance) people might have issue (sharing for people on k8s)

We will upgrade to the latest version (2.6.0) or the same version you seemed to be using on production on your end 2.5.0 or 2.5.1 https://github.com/getsentry/onpremise/blob/master/docker-compose.yml#L82

Thank you once again for your time and sorry for the trouble.

**Comment by beezz** [comment]
Thank you @Sawk for your investigation and write up. Would be amazing if you can confirm that upgrading Kafka solves the issue for you.

I will close the issue now but please feel free to reopen it in case you find out that the problem is elsewhere.
